store/mongo/datastore_mongo.go

```diff
@@ -368,22 +375,47 @@ func (db *DataStoreMongo) upsertAttributes(
 	case 1:
 		var res *mongo.UpdateResult
 		if withRevision {
-			filter = bson.M{
-				"_id":         devices[0].Id,
-				DbDevRevision: bson.M{"$lt": devices[0].Revision},
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+					etagField:     bson.M{"$exists": true, "$eq": *etag},
+				}
+				if err := checkMatchedCount(ctx, c, filter); err != nil {
+					return &model.UpdateResult{MatchedCount: 0}, nil
+				}
+
+			} else {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+				}
 			}
 			update[DbDevRevision] = devices[0].Revision
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
-			}
 		} else {
-			filter = map[string]interface{}{"_id": devices[0].Id}
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":     devices[0].Id,
+					etagField: bson.M{"$exists": true, "$eq": *etag},
```

Comment:
$eq is enough, no need to use $exists as well

Reply:
Makes sense